plugins/chat/app/services/chat/auto_leave_channels.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+module Chat
+  class AutoLeaveChannels
+    include Service::Base
+
+    ALLOWED_GROUP_PERMISSIONS ||= [
+      CategoryGroup.permission_types[:create_post],
+      CategoryGroup.permission_types[:full],
+    ]
+
+    policy :chat_enabled?
+
+    contract do
+      attribute :event
+      attribute :user_id, :integer
+      attribute :group_id, :integer
+      attribute :channel_id, :integer
+      attribute :category_id, :integer
+    end
+
+    step :remove_memberships
+
+    private
+
+    def chat_enabled?
+      SiteSetting.chat_enabled
+    end
+
+    def remove_memberships
+      group_ids = SiteSetting.chat_allowed_groups_map
+      group_permissions = ALLOWED_GROUP_PERMISSIONS
+
+      if !group_ids.include?(Group::AUTO_GROUPS[:everyone])
+        sql = <<~SQL
+          DELETE FROM user_chat_channel_memberships uccm
+          WHERE NOT EXISTS (
+            SELECT 1 
+            FROM group_users gu 
+            WHERE gu.user_id = uccm.user_id 
+            AND gu.group_id IN (:group_ids)
+          )
```

Comment:
This is going to delete all the chat memberships (including DMs) for users who don't have access to chat anymore (via the `chat_allowed_groups` site setting).

Reply:
Previously we didn't delete direct message channels, should we keep existing functionality here? Deleting them would mean that users would have to manually join their DM channels again if the group was once again added to `chat_allowed_groups`.

An alternative option might be to set `following` to false, that way memberships could be restored if the group was later added and they wouldn't lose their chat history (how I understand it).

Reply:
After some discussion about the tradeoffs, deleting records feels much safer, easier, and performant just to avoid people having to get invited by into group DMs (the other chat memberships will be either automagically re-created or are very easy enough to do so (for 1-1 DMs)).